### PR TITLE
Refactor `maybe_create_object_store_from_uri`

### PR DESCRIPTION
**Pull request by irenedea**
# What does this PR do?
Refactors `maybe_create_object_store_from_uri` to use a dictionary, which maps backend strings to object store factory functions.

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/utils/file_helpers.py [hunk 1]
```diff
@@ -393,13 +393,41 @@ def parse_uri(uri: str) -> tuple[str, str, str]:
         return backend, bucket_name, path.lstrip('/')
 
 
-def maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
-    """Automatically creates an :class:`composer.utils.ObjectStore` from supported URI formats.
+# Registry for object store creation functions
+object_store_registry: dict[str, Callable[[str, str], ObjectStore]] = {}
```
Comment (irenedea) [inline composer/utils/file_helpers.py:397]: Would this be nicer encapsulated in the ObjectStore class as a global variable there? 
Pros: encapsulation, Cons: it's further away from its main and only usage point `maybe_create_object_store_from_uri`

File: composer/utils/file_helpers.py [hunk 2]
```diff
@@ -408,54 +424,38 @@ def maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
...
+    # If backend is empty, assume local path and return None
     if backend == '':
         return None
-    if backend == 's3':
-        return S3ObjectStore(bucket=bucket_name)
-    elif backend == 'wandb':
+
+    # Check if backend is registered
+    if backend in BACKEND_TO_OBJECT_STORE_FACTORY:
+        return BACKEND_TO_OBJECT_STORE_FACTORY[backend](bucket_name, path)
```
Comment (mvpatel2000) [inline composer/utils/file_helpers.py:434]: minor nit: can we make this part of the if elif flow?

**Review by mvpatel2000 — COMMENTED** [review]
Should we add a test for registering a mock option?

File: composer/utils/file_helpers.py (hunk 2 above)
Comment (irenedea) [reply composer/utils/file_helpers.py]: done

**Comment by irenedea** [comment]
@mvpatel2000 The existing test `test_maybe_create_object_store_from_uri` already tests that items in the dictionary are read from. I think that should be sufficient to handle the case where someone extends the dictionary.

File: composer/utils/file_helpers.py [hunk 3]
```diff
@@ -408,54 +424,38 @@ def maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
...
-            secret_environ='AZURE_ACCOUNT_ACCESS_KEY',
+            f'There is no implementation for WandB load_object_store via URI. Please use WandBLogger',
         )
     elif backend == 'dbfs':
         if path.startswith(MLFLOW_DBFS_PATH_PREFIX):
             store = None
             if dist.get_global_rank() == 0:
                 store = MLFlowObjectStore(path)
-
-                # The path may have had placeholders, so update it with the experiment/run IDs initialized by the store
```
Comment (mvpatel2000) [inline composer/utils/file_helpers.py:437]: why are comments removed?

**Review by mvpatel2000 — APPROVED** [review]
LGTM. not sure why comments are removed though

File: composer/utils/file_helpers.py (hunk 3 above)
Comment (irenedea) [reply composer/utils/file_helpers.py:437]: AI coding assistants 😛 , will add back in...

**Pull request merged**